src/fvtpf/tdyfvtpf_snes.c

```diff
@@ -0,0 +1,714 @@
+    PetscInt cell_id_dn = cell_ids[1];
+
+    PetscReal Res;
+    ierr = RichardsResidual(fvtpf, tdy->dm, tdy->matprop, iface, &Res);
+
+    if (cell_id_up >= 0 && cells->is_local[cell_id_up]) {
+      r_ptr[cell_id_up] += Res;
+    }
+
+    if (cell_id_dn >= 0 && cells->is_local[cell_id_dn]) {
+      r_ptr[cell_id_dn] -= Res;
+    }
+
+  }
+
+  // Compute contribution to residual for boundary faces
+  for (PetscInt iface=0; iface<mesh->num_faces; iface++) {
+
+    if (!faces->is_local[iface]) continue; // skip non-local face
+    if (faces->is_internal[iface]) continue; // skip internal faces
+    if (faces->bc_type[iface] == NEUMANN_BC) continue; // skip non-flux faces
+
+    PetscInt *cell_ids, num_face_cells;
+    ierr = TDyMeshGetFaceCells(mesh, iface, &cell_ids, &num_face_cells); CHKERRQ(ierr);
+
+    PetscReal Res;
+    switch (faces->bc_type[iface]) {
+    case NEUMANN_BC:
+      break;
+    case DIRICHLET_BC:
+      ierr = RichardsBCResidual(fvtpf, tdy->dm, tdy->matprop, iface, &Res);
+      break;
+    case SEEPAGE_BC:
+      ierr = RichardsSeepageBCResidual(fvtpf, tdy->dm, tdy->matprop, iface, &Res);
+      break;
+    }
+
+    PetscInt cell_id_dn = cell_ids[1];
+    r_ptr[cell_id_dn] -= Res;
+  }
+
+  PetscReal accum_current;
+  for (PetscInt icell = 0; icell<mesh->num_cells; icell++) {
+
+    if (!cells->is_local[icell]) continue;
+
+    ierr = TDyFVTPFSNESAccumulation(tdy,icell,&accum_current); CHKERRQ(ierr);
+
+    r_ptr[icell] += accum_current - accum_prev[icell];
+    r_ptr[icell] -= fvtpf->source_sink[icell];
+
+  }
+
+  ierr = VecRestoreArray(R, &r_ptr); CHKERRQ(ierr);
+  ierr = VecRestoreArray(tdy->accumulation_prev, &accum_prev); CHKERRQ(ierr);
+
+  TDY_STOP_FUNCTION_TIMER()
+
+  PetscFunctionReturn(0);
+}
+
+/* -------------------------------------------------------------------------- */
+PetscErrorCode TDyFVTPFSNESJacobian(SNES snes,Vec U,Mat A, Mat B,void *ctx) {
+
+  TDy      tdy = (TDy)ctx;
+  TDyFVTPF *fvtpf = tdy->context;
+  TDyMesh  *mesh = fvtpf->mesh;
+  TDyCell  *cells = &mesh->cells;
+  TDyFace  * faces = &mesh->faces;
+
+  PetscErrorCode ierr;
+
+  PetscFunctionBegin;
+  TDY_START_FUNCTION_TIMER()
+
+  ierr = MatZeroEntries(B); CHKERRQ(ierr);
+
+  // Compute contribution to Jacobian for internal faces
+  for (PetscInt iface=0; iface<mesh->num_faces; iface++) {
+
+    if (!faces->is_local[iface]) continue; // skip non-local face
+    if (!faces->is_internal[iface]) continue; // skip boundary faces
+
+    PetscInt *cell_ids, num_face_cells;
+    ierr = TDyMeshGetFaceCells(mesh, iface, &cell_ids, &num_face_cells); CHKERRQ(ierr);
+    PetscInt cell_id_up = cell_ids[0];
+    PetscInt cell_id_dn = cell_ids[1];
+
+    PetscReal Jup, Jdn;
+    ierr = RichardsJacobian(fvtpf, tdy->dm, tdy->matprop, iface, &Jup, &Jdn);
+
+    if (cell_id_up >= 0 && cells->is_local[cell_id_up]) {
+      //r_ptr[cell_id_up] += Res;
+      ierr = MatSetValuesLocal(B,1,&cell_id_up,1,&cell_id_up,&Jup,ADD_VALUES);CHKERRQ(ierr);
```

Comment:
These could be vectorized to improve performance. Not critical for this PR, though.